File: typescript/src/clients/openresponses.ts
// =============================================================================
// OpenResponses Client — HTTP client for /v3/compat/openresponses
// =============================================================================

import type { ORRequest, ORResponse, ORStreamEvent } from "../agent/types.js";
import { BaseClient } from "../client-base.js";
import type { RequestOptions } from "../types.js";

const ENDPOINT = "/v3/compat/openresponses";

/**
 * Client for the OpenResponses endpoint.
 *
 * Provides two methods:
 * - `create()` — synchronous (stream: false), returns ORResponse
 * - `createStream()` — streaming (stream: true), yields ORStreamEvent
 */
export class OpenResponsesClient extends BaseClient {
  /**
   * Create a response (non-streaming).
   * Sends the request with `stream: false` and returns the complete ORResponse.
   */
  async create(request: ORRequest, options?: RequestOptions): Promise<ORResponse> {
    return this._post<ORResponse>(ENDPOINT, { ...request, stream: false }, options);
  }

  /**
   * Create a streaming response.
   * Sends the request with `stream: true` and yields ORStreamEvent objects
   * parsed from the SSE stream.
   *
   * OpenResponses SSE format uses named event types:
   * ```
   * event: response.output_text.delta
   * data: {"type":"response.output_text.delta","output_index":0,"delta":"Hello"}
   * ```
   */
  async *createStream(
    request: ORRequest,
    options?: RequestOptions,
  ): AsyncGenerator<ORStreamEvent, void, undefined> {
    // Use fetchRaw for unified error mapping (4xx/5xx → typed subclasses).
    const response = await this.fetchRaw(
      `${this.baseUrl}${ENDPOINT}`,
      {
        method: "POST",
        body: JSON.stringify({ ...request, stream: true }),
      },
      {
        ...options,
        headers: { ...options?.headers, Accept: "text/event-stream" },
      },
    );

    if (!response.body) {
      return;
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });

        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          const trimmed = line.trim();

          // Empty lines and comments — skip
          if (!trimmed || trimmed.startsWith(":")) {
            continue;
          }

          // event: lines — skip (type is embedded in the data JSON)
          if (trimmed.startsWith("event:")) {
            continue;
          }

          // data: lines — parse and yield
          if (trimmed.startsWith("data:")) {
            const data = trimmed.slice(5).trim();

            if (data === "[DONE]") {
              return;
            }

            if (!data) {
              continue;
            }

            try {
              const parsed = JSON.parse(data) as ORStreamEvent;
              yield parsed;
            } catch {
              // Skip malformed JSON
            }
          }
        }
      }

      // Process remaining buffer
      if (buffer.trim()) {
        const trimmed = buffer.trim();
        if (trimmed.startsWith("data:")) {
          const data = trimmed.slice(5).trim();
          if (data && data !== "[DONE]") {
            try {
              yield JSON.parse(data) as ORStreamEvent;
            } catch {
              // Skip malformed JSON
            }
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}

File: python/src/opperai/clients/openresponses.py
"""OpenResponses client — HTTP client for POST /v3/compat/openresponses."""

from __future__ import annotations

from collections.abc import AsyncIterator, Iterator
from typing import Any

from .._base_client import BaseClient
from ..types import RequestOptions

ENDPOINT = "/v3/compat/openresponses"


class OpenResponsesClient:
    """Client for the OpenResponses endpoint.

    Provides four methods:
    - ``create()`` / ``create_async()`` — non-streaming, returns the full response dict
    - ``create_stream()`` / ``create_stream_async()`` — streaming, yields raw SSE event dicts
    """

    def __init__(self, client: BaseClient) -> None:
        self._client = client

    # --- Non-streaming --------------------------------------------------------

    def create(
        self,
        request: dict[str, Any],
        options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        """Send a non-streaming request. Returns the complete ORResponse dict."""
        return self._client._post(
            ENDPOINT,
            body={**request, "stream": False},
            options=options,
        )

    async def create_async(
        self,
        request: dict[str, Any],
        options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        """Send a non-streaming request (async). Returns the complete ORResponse dict."""
        return await self._client._post_async(
            ENDPOINT,
            body={**request, "stream": False},
            options=options,
        )

    # --- Streaming ------------------------------------------------------------

    def create_stream(
        self,
        request: dict[str, Any],
        options: RequestOptions | None = None,
    ) -> Iterator[dict[str, Any]]:
        """Send a streaming request. Yields raw SSE event dicts.

        The event type is in the ``type`` field of each dict, e.g.
        ``"response.output_text.delta"``, ``"response.completed"``, etc.
        """
        return self._client._stream_sse_raw(
            ENDPOINT,
            body={**request, "stream": True},
            options=options,
        )

    async def create_stream_async(
        self,
        request: dict[str, Any],
        options: RequestOptions | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Send a streaming request (async). Yields raw SSE event dicts."""
        return self._client._stream_sse_raw_async(
            ENDPOINT,
            body={**request, "stream": True},
            options=options,
        )
